### PR TITLE
Retry when starting transactions times out

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
We fail without retrying when attempting to start transactions. This is most obvious around leader elections.

**Implementation Description (bullets)**:
* Add a `TimeoutHandler` as a configurable option on the autobatcher, as well as a new exception type for the start transaction batcher to use (since TFRE isn't accessible here!)

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added test to AutobatchersTests.

**Concerns (what feedback would you like?)**:
Should we be retrying here?

**Where should we start reviewing?**:
Diff

**Priority (whenever / two weeks / yesterday)**:
Whenever.

### CODE REVIEW

**Review by sudiksha27 — COMMENTED** [review]
Few comments but looks good otherwise!

File: lock-api/src/main/java/com/palantir/lock/client/StartTimestampFailedException.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+public final class StartTimestampFailedException extends RuntimeException {
+    public StartTimestampFailedException(String message, Throwable cause) {
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/StartTimestampFailedException.java:20]: agreed - renamed.

File: lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java [hunk 2]
```diff
@@ -48,6 +48,8 @@ static BatchingIdentifiedAtlasDbTransactionStarter create(LockLeaseService lockL
                 Autobatchers.independent(consumer(lockLeaseService, cache))
                         .safeLoggablePurpose("transaction-starter")
                         .batchFunctionTimeout(Duration.ofSeconds(30))
+                        .timeoutHandler(exception -> new StartTimestampFailedException(
+                                "Timed out while attempting to start transactions", exception))
                         .build();
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:53]: Ah, we absolutely do! My bad here.

[the remark this answers is not in the document]
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:53]: is this done?

[the remark this answers is not in the document]
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:53]: Should be - is it not in the diff?

[the remark this answers is not in the document]
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:53]: Should be - is it not in the diff?

**Review by sudiksha27 — APPROVED** [review]
LGTM!

File: lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:53]: fooled by commit messages 😵‍💫

**Pull request merged**